File: 1352-product_of_the_last_k_numbers/go/main.go
package main

import (
	"fmt"
)

type ProductOfNumbers struct {
	products []int
}

func Constructor() ProductOfNumbers {
	return ProductOfNumbers{
		products: []int{},
	}
}

func (this *ProductOfNumbers) Add(num int) {
	if num == 0 {
		this.products = []int{}
	} else {
		for i := 0; i < len(this.products); i++ {
			this.products[i] = this.products[i] * num
		}
		this.products = append(this.products, num)
	}
}

func (this *ProductOfNumbers) GetProduct(k int) int {
	if k > len(this.products) {
		return 0
	}
	return this.products[len(this.products)-k]
}

func main() {
	obj := Constructor()
	obj.Add(3)
	obj.Add(0)
	obj.Add(2)
	obj.Add(5)
	obj.Add(4)
	if val := obj.GetProduct(2); val != 20 {
		fmt.Printf("1. Test failed, expected 20, got %d", val)
		return
	}
	if val := obj.GetProduct(3); val != 40 {
		fmt.Printf("2. Test failed, expected 40, got %d", val)
		return
	}
	if val := obj.GetProduct(4); val != 0 {
		fmt.Printf("3. Test failed, expected 0, got %d", val)
		return
	}
	obj.Add(8)
	if val := obj.GetProduct(2); val != 32 {
		fmt.Printf("4. Test failed, expected 32, got %d", val)
		return
	}
	fmt.Println("Passed")
}

File: 1352-product_of_the_last_k_numbers/ruby/main.rb
class ProductOfNumbers
  def initialize
    @products = {}
  end

=begin
    :type num: Integer
    :rtype: Void
=end
  def add(num)
    if num == 0
      @products = {}
    else
      @products.each do |key, value|
        @products[key] = value * num
      end
      @products[@products.length] = num
    end
  end

=begin
    :type k: Integer
    :rtype: Integer
=end
  def get_product(k)
    return 0 if k > @products.length
    @products[@products.length - k]
  end
end

if __FILE__ == $0
  obj = ProductOfNumbers.new

  obj.add(3)
  obj.add(0)
  obj.add(2)
  obj.add(5)
  obj.add(4)

  if obj.get_product(2) != 20
    puts "1. Test failed, expected 20"
    exit
  end

  if obj.get_product(3) != 40
    puts "2. Test failed, expected 40"
    exit
  end

  if obj.get_product(4) != 0
    puts "3. Test failed, expected 0"
    exit
  end

  obj.add(8)

  if obj.get_product(2) != 32
    puts "4. Test failed, expected 32"
    exit
  end

  puts "Passed"
end

